build: convert CLDR locale extraction from Gulp to Bazel tool

devversion opened the pull request:
See individual commits.

This PR initially contained the update to CLDR 39 too, but I've decided to switch back to CLDR 37 so that
we can land this (without having to wait for a new major). I've also compared old generated files with the new
generated files to make sure nothing is missing/changed in terms of the locale data.

devversion committed:
refactor: remove checked-in locale files

This is a pre-refactor commit allowing us to move
the CLDR locale generation to Bazel where files would
no longer be checked-in, except for the `closure-locale`
file that is synced into Google3.

devversion committed:
build: convert CLDR locale extraction from Gulp to Bazel tool

Converts the CLDR locale extraction script to a Bazel tool.
This allows us to generate locale files within Bazel, so that
locales don't need to live as sources within the repo. Also
it allows us to get rid of the legacy Gulp tooling.

The migration of the Gulp script to a Bazel tool involved the
following things:

  1. Basic conversion of the `extract.js` script to TypeScript.
     This mostly was about adding explicit types. e.g. adding `locale:
     string` or `localeData: CldrStatic`.

  2. Split-up into separate files. Instead of keeping the large
     `extract.js` file, the tool has been split into separate files.
     The logic remains the same, just that code is more readable and
     maintainable.

  3. Introduction of a new `index.ts` file that is the entry-point
     for the Bazel tool. Previously the Gulp tool just generated
     all locale files, the default locale and base currency files
     at once. The new entry-point accepts a mode to be passed as
     first process argument. based on that argument, either locales
     are generated into a specified directory, or the default locale,
     base currencies or closure file is generated.

     This allows us to generate files with a Bazel genrule where
     we simply run the tool and specify the outputs. Note: It's
     necessary to have multiple modes because files live in separate
     locations. e.g. the default locale in `@angular/core`, but the
     rest in `@angular/common`.

  4. Removal of the `cldr-data-downloader` and custom CLDR resolution
     logic. Within Bazel we cannot run a downloader using network.

     We switch this to something more Bazel idiomatic with better
     caching. For this a new repository rule is introduced that
     downloads the CLDR JSON repository and extracts it. Within
     that rule we determine the supported locales so that they
     can be used to pre-declare outputs (for the locales) within
     Bazel analysis phase. This allows us to add the generated locale
     files to a `ts_library` (which we want to have for better testing,
     and consistent JS transpilation).

     Note that the removal of `cldr-data-downloader` also requires us to
     add logic for detecting locales without data. The CLDR data
     downloader overwrote the `availableLocales.json` file with a file
     that only lists locales that CLDR provides data for. We use the
     official `availableLocales` file CLDR provides, but filter out
     locales for which no data is available. This is needed until we
     update to CLDR 39 where data is available for all such locales
     listed in `availableLocales.json`.

devversion committed:
build: wire up new CLDR generation tool within Bazel

Introduces a few Starlark macros for running the new Bazel
CLDR generation tool. Wires up the new tool so that locales
are generated properly. Also updates the existing
`closure-locale` file to match the new output generated by the Bazel tool.

This commit also re-adds a few locale files that aren't
generated by CLDR 37, but have been accidentally left in
the repository as the Gulp script never removed old locales
from previous CLDR versions. This problem is solved with the
Bazel generation of locale files, but for now we re-add these
old CLDR 33 locale files to not break developers relying on these
(even though the locale data indicies are incorrect; but there might
be users accessing the data directly)

devversion committed:
build: build locale files for legacy saucelabs job

Given that the locale files are now generated through
Bazel, the files are no longer checked-in and the
legacy TSC compilation fails due to imports resolving
to non-existent files. We fix this for the legacy
saucelabs job by copying the generated TS files into
the sources (which is acceptable for the isolated CI job)

petebacondarwin commented on packages/common/locales/generate-locales-tool/index.ts:
```diff
@@ -0,0 +1,83 @@
+import {generateLocale} from './locale-file';
+import {generateLocaleGlobalFile} from './locale-global-file';
+
+async function main() {
```
Could you document the command line arguments that this file accepts?

petebacondarwin replied:
Actually I feel like it would actually be cleaner to just have a different executable for each of the modes, rather than one that handles them all. Each mode is pretty independent of the others, and now that you have nicely refactored the code that actually does the work, there is no real need to keep them all in the same `main()`.  This would also make each mode more self-documenting, perhaps?

devversion replied:
I've added a new folder called `bin` that exposes the tool entry-points. I didn't add the `main` functions to existing files I've split-up because that would make them less readable IMO and having a separate `bin` folder is more clear IMO + it's easier to have a place for defining the `BASE_LOCALE` (which is `en`)

petebacondarwin commented on packages/common/locales/generate-locales-tool/index.ts:
```diff
@@ -0,0 +1,83 @@
+  const baseLocaleData = getLocaleData('en');
+  const baseCurrencies = generateBaseCurrencies(baseLocaleData, false);
+
+  const [mode, outputDir] = process.argv.slice(2);
```
Could you grab the `mode` and `outputDir` from the argv in the block at the bottom of this file, and then pass them into the `main()` function as arguments? E.g.

```ts
if (require.main === module) {
  const [mode, outputDir] = process.argv.slice(2);
  main(mode, outputDir);
}
```

petebacondarwin commented on packages/common/locales/generate-locales-tool/license-header.ts:
```diff
@@ -0,0 +1,15 @@
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */`;
```
Add a newline to the end of the file for consistency?

devversion replied:
done. we should start linting this! 😄

petebacondarwin commented on packages/common/locales/generate-locales-tool/object-stringify.ts:
```diff
@@ -0,0 +1,27 @@
+ * Stringifies the given object while preserving `undefined` values which would usually
+ * be transformed into `null` with JSON5. We want to preserve `undefined` because in generated
+ * JavaScript, the `undefined` values are separate from `null`, and `undefined` can be minified
+ * more efficiently. For example in arrays: `[, , someValue]`.
```
I'm not sure that we do minify in this way any more, since instead we use `const u = undefined` and add `u` everywhere. I think there was a problem (probably with IE11) with handling "empty" elements in an array.

petebacondarwin replied:
But that is probably for another PR.

devversion replied:
Yeah, you are referring to the `u` deduping case. I think compared to `null` though, `undefined` can be minified better as it allows us to just add the coma delimiter between elements, while `null` would need to be explicitly set.

petebacondarwin replied:
I believe that at some point we generated with empty elements but then switched to the `u` approach due to some browsers complaining. I don't think that our current minification will replace the `u` identifiers with empty array elements.

petebacondarwin commented on packages/common/locales/BUILD.bazel:
```diff
@@ -1,18 +1,43 @@
 package(default_visibility = ["//visibility:public"])
 
+generate_closure_locales_file(
+    name = "closure_locales_file_generated",
```
OOC why does this rule name end with `_generated` while the `base_currencies_file` does not?

devversion replied:
I did this mainly to disambiguate from the target that ensures the checked-in file is up-to-date.

petebacondarwin commented on packages/common/locales/BUILD.bazel:
```diff
@@ -1,18 +1,43 @@
     name = "locales",
-    srcs = glob(
-        ["**/*.ts"],
+    # TODO(devversion): Remove glob for checked-in legacy locale files that haven't been
+    # removed in the past (when CLDR has been updated). These can be removed in a major.
```
I would recommend making the PR for this now, but just marking it as `target: major` so that it will not be forgotten when we open up 13.x.0 for commits.

devversion replied:
@petebacondarwin Are you saying I should move this PR to CLDR 39 already? I did that initially but went back to CLDR 37 so that we can land the infrastructure rework without having to wait for a new major.

petebacondarwin replied:
Oh no, I just meant that once this PR lands we should not wait to create the next PR marking it to target the next major version.

devversion replied:
For sure! Agreed 😄

petebacondarwin commented on packages/common/locales/closure-locale.ts:
```diff
@@ -6,6228 +6,1454 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-// THIS CODE IS GENERATED - DO NOT MODIFY
```
I think it is valuable to keep this comment in the file, given that it is still generated.
You would not believe the number of times people try to change such files in PRs.

petebacondarwin reviewed (commented):
Fantastic work @devversion! I love the way you can skip from TS to Skylark and back so fluently. Here is my initial feedback. No serious blockers that I can see. Mainly just a bunch of NITs and suggestions.

devversion commented on package.json:
```diff
@@ -164,17 +164,18 @@
     "clang-format": "^1.4.0",
-    "cldr": "7.0.0",
-    "cldr-data-downloader": "^0.3.5",
+    "cldr": "5.7.0",
```
Good question. The reason is that we recently updated the `cldr` package through renovate a couple of times. This is wrong because it updated us to CLDR 39. This downgrade gets us back to CLDR 37.

We can update back to CLDR 7 once we update to CLDR 39. I've landed some fixes in `cldr` package that allow us to generate plural functions w/ CLDR 39.

devversion commented on packages/common/locales/generate-locales-tool/cldr-data.ts:
```diff
@@ -0,0 +1,77 @@
+// TODO: See if we can improve this by having better types for `cldrjs`.
+const cldrjs: typeof import('cldrjs') = require('cldrjs');
+
+/** Path to the CLDR data repository. */
```
Yeah, it is. Repository is the term within Bazel. e.g. `@cldr_rule` is a repository.

devversion replied:
I specifically mention Bazel in the comment now. that might help.

devversion commented on packages/common/locales/generate-locales-tool/cldr-data.ts:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Fetches necessary CLDR data from the specified CLDR directory.
+ */
+function fetchLocaleDataFromCldr(cldrDataDir: string): object[] {
```
Agreed 👍

devversion commented on packages/common/locales/generate-locales-tool/array-deduplication.ts:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ * To create smaller locale files, we remove duplicated data.
+ * To be make this work we need to store similar data in arrays, if some value in an array
+ * is undefined, we can take the previous defined value instead, because it means that it has
+ * been deduplicated.
```
Yeah, this is not really part of the PR, but I'm changing it while being at it 😄 This stuff needs more cleanup in the future.

devversion commented on packages/common/locales/generate-locales-tool/cldr-data.ts:
```diff
@@ -0,0 +1,77 @@
+  cldrjs.load(...localeData as [object, ...object[]]);
+}
+
+export function getLocaleData(localeName: string): CldrStatic {
+  return new cldrjs(localeName);
+}
```
I like the idea of having a class 👍

devversion commented on packages/common/locales/generate-locales-tool/array-deduplication.ts:
```diff
@@ -0,0 +1,38 @@
+      dedup.push(undefined);
+    }
+  }
+  return dedup;
```
Great idea. At first glance it looks like this could work. We might want to do that in a next major in case people rely on the length.

devversion replied:
Created https://github.com/angular/angular/issues/42315 for tracking.

devversion commented:
@petebacondarwin Thanks for reviewing so thoroughly! I've addressed the feedback. While doing the requested cleanup of a few things within the old `extract.js` script, I also went ahead and simplified the base currency generation a bit.

Previously, we generated two objects of base currencies. Once for the `currencies.ts` file in `common/src/i18n`, and one just for comparisons. I've changed it so that we only have to generate one type of `BaseCurrencies` object. The old `extract.js` script generated the two different objects so that it could do comparisons using `toString()` (the locale currencies do not contain the fraction digits, while the base currencies does; -> this is the reason for the `addDigits` flag in `generateBaseCurrencies`)

petebacondarwin commented on packages/common/locales/generate-locales-tool/bin/BUILD.bazel:
```diff
@@ -0,0 +1,33 @@
+[nodejs_binary(
+    name = entrypoint,
+    data = [
+        "//packages/common/locales/generate-locales-tool/bin",
```
Could this just be `":bin"`?

devversion replied:
`bin` is not a target within `generate-locales-tool`, so unfortunately not 😄

petebacondarwin commented on packages/common/locales/generate-locales-tool/bin/write-locale-files-to-dist.ts:
```diff
@@ -0,0 +1,69 @@
+
+
+if (require.main === module) {
+  const [outputDir] = process.argv.slice(2);
```
NIT: I know this is a standard style but it is a bit overkill when you could just do:

```ts
const outputDir = process.argv[2];
```

petebacondarwin replied:
Actually is this still `argv[2]` and not `argv[1]`?

devversion replied:
Yeah, because the command usually is `node <script> <args>` (at least in Bazel), correct me if I'm wrong 😄

petebacondarwin commented on packages/common/locales/generate-locales-tool/object-stringify.ts:
```diff
@@ -0,0 +1,27 @@
+  UNDEFINED_PLACEHOLDER_REGEX.lastIndex = 0;
+
+  return result.replace(UNDEFINED_PLACEHOLDER_REGEX, 'undefined');
+}
```
😱

petebacondarwin reviewed (approved):
Great work! I really like the changes you made to the binaries.

AndrewKushnir reviewed (approved):
@devversion looks awesome! 👍 

Thanks for documenting the logic, this is really helpful. While you have all the context, I was thinking that it might also be beneficial to add a README.md file (into the `packages/common/locales/generate-locales-tool` folder) to describe some aspects of the process and optmiizations from high-level perspective (while keeping the details in the code as well).

devversion committed:
build: add documentation for `generate-locales-tool`

The CLDR extraction tool has been reworked to run as part of Bazel.
This adds a initial readme explaining what the tool generates. It's
far from a detailed description but it can serve as foundation for more
detailed explanations.

devversion commented:
@petebacondarwin Thanks! Addressed the feedback.
@AndrewKushnir I've added a little README. It's far from being very detailed, but might make some things more clear (and can serve as a foundation). Please have another look!

AndrewKushnir commented:
@devversion great, thanks for adding the readme file 👍

josephperrott commented on packages/common/locales/generate-locales-tool/bin/write-locale-files-to-dist.ts:
```diff
@@ -0,0 +1,71 @@
+if (require.main === module) {
+  // The first argument is expected to be a path resolving to a directory
+  // where all locales should be generated into.
+  const outputDir = process.argv[2];
```
Should we `require()` the `outputDir` to make sure it exists?

devversion replied:
Yeah, we can add a check.

petebacondarwin replied:
> Should we require() the outputDir to make sure it exists?

What does it mean to `require()` this path? Wouldn't that try to load a `package.json` from this path? If the path doesnt exist then the `writeFileSync` above will fail pretty clearly.

devversion replied:
Yeah, I assumed Joey was referring to something like `existsSync`

devversion replied:
Agreed that the standard Node error would be sufficient too.

josephperrott commented on packages/common/locales/generate-locales-tool/cldr-data.ts:
```diff
@@ -0,0 +1,93 @@
+
+  /** Gets a list of all locales CLDR provides data for. */
+  private _getAvailableLocales(): string[] {
+    return require(`${this.cldrDataDir}/cldr-core-37.0.0/availableLocales.json`)
```
Can the path of the availableLocales be extracted out of the class? Just to colocate it with the other version specific values.

devversion replied:
Unfortunately not in a reasonable way. The data is stored in `Cldr._resolved` and we'd need to specify a locale to get access (which I think is odd given this is global data not specific to a given locale)

devversion replied:
Ah, I was thinking you want me to extract the data out of the `cldrjs` package where the data is already populated in. Created a constant for the path. 👍

josephperrott commented on packages/common/locales/generate-locales-tool/closure-locale-file.ts:
```diff
@@ -0,0 +1,158 @@
+  return `${licenseHeader}
+
+// THIS CODE IS GENERATED - DO NOT MODIFY. See "BUILD.bazel" for details.
+// clang-format off
```
We should remove this comment as we aren't committing it so we won't run a formatter.

Additionally, we should ignore formatting using the formatters ignore support rather than opt out here.

devversion replied:
This file is still checked-in as it's the one for google3. Agreed on ignoring the formatting in the config though.

devversion replied:
Updated.

josephperrott commented on packages/common/locales/generate-locales-tool/day-periods.ts:
```diff
@@ -0,0 +1,85 @@
+    });
+  }
+
+  // Implements inheritance for the day periods. As of CLDR v39, the `nn` locale
```
Since we aren't using v39 yet, do we need to do something differently here?

devversion replied:
Oh, this is a leftover from my initial PR working with CLDR 39. Are we good keeping this for now. Happy to also remove it. It shouldn't result in any changes as this just serves as a fallback we'd not hit in CLDR 37.

josephperrott replied:
I am happy to have it stay there, I just wanted to make sure it was not going to cause issue.

josephperrott commented on packages/common/locales/generate-locales-tool/license-header.ts:
```diff
@@ -0,0 +1,15 @@
+/**
```
I love how we  have the license just about the licenseHeader constant.

devversion commented on packages/common/BUILD.bazel:
```diff
@@ -1,10 +1,16 @@
 
 package(default_visibility = ["//visibility:public"])
 
+generate_base_currencies_file(
```
I don't see a reason why this would be necessary. It's better to not check-in generated files as it simplifies the process of updating CLDR and also avoids adding another Bazel target for testing the checked-in file.

josephperrott replied:
My concern was about how we will have non-generated files which import the generated file.  And I am worried that its going to make intellisense go crazy

devversion replied:
Yeah, but that would apply for all other locale files too (some are imported in tests). I don't think this is worth doing. It's true that the IDE will show such imports as failing, but I consider that as acceptable given the benefits of using just a `genrule`. People don't deal a lot with locales either, and you get proper error diagnostics when building anyway.

devversion commented on packages/common/locales/generate-locales-tool/README.md:
```diff
@@ -0,0 +1,64 @@
+# generate-locales-tool
```
It's true that this contains information about how the locale files are structured, but at the same time it contains information about the optimizations performed by the tool (for these to understand it's necessary to explain how the locales are structured).

josephperrott reviewed (approved):
LGTM

devversion commented:
@josephperrott Updated this PR to check-in the `locale_en.ts` and `currencies.ts` file. Can you please run another presubmit?

devversion committed:
build: simplify generation of closure locale file

In the past, the closure file has been generated so that all individual
locale files were imported individually. This resulted in a huge
slow-down in g3 due to the large amount of imports.

With 90bd984ff74f7605d7c08fd9fdbf610ba7fa67a5 this changed so that we
inline the locale data for the g3 closure locale file. Also the file
only contained data for locales being supported by Closure. For this a
list of locales has been extracted from Closure Compiler, as well as a
list of locale aliases.

This logic is prone to CLDR version updates, and also broke as part of
the Gulp -> Bazel migration where this logic has been slightly modified
but caused issues in G3. e.g. a locale `zh-Hant` was requested in g3,
but the locale data had the name of the alias locale that provided the
data at index zero (which represents the locale name). Note that the
locale names at index zero always could differentiate from the requested
`goog.LOCALE` due to the aliasing logic. This just didn't come up before.

We simplify this logic by generating a `goog.LOCALE` case for all
locales CLDR provides data for. We don't need to bother about aliasing
because with the refactorings to the CLDR generation tool, all locales
are built (which also captures the aliases), and we can generate the locale
file on the fly (which has not been done before).

jessicajaniuk commented:
This unfortunately broke targets inside Google. I'll be pushing revert commits shortly.

jessicajaniuk commented:
The revert commits are in PR 42521.

devversion committed:
build: generate alias locale data for closure locale

Within Google, closure compiler is used for dealing with translations.
We generate a closure-compatible locale file that allows for
registration within Angular, so that Closure i18n works well together
with Angular applications. Closure compiler does not limit its
locales to BCP47-canonical locale identifiers. This commit updates
the generation logic so that we also support deprecated (but aliased)
locale identifiers, or other aliases which are likely used within
Closure. We use CLDR's alias supplemental data for this. It instructs
us to alias `iw` to `he` for example. `iw` is still supported in Closure.

Note that we do not manually extract all locales supported in Closure;
instead we only support the CLDR canonical locales (as done before) +
common aliases that CLDR provides data for. We are not aware of other
locale aliases within Closure that wouldn't be part of the CLDR aliases.
If there would be, then Angular/Closure would fail accordingly.

jessicajaniuk commented:
The changes look good to me. I've ran a regular presumbit, and that went green. I'm going to run a TGP tonight to fully verify the safety of this PR.

devversion commented:
FYI: I've re-tested the locale data files (which are shipped as part of `@angular/common/locales`) to ensure the data matches with the current data in Angular `master` (in case the tool would have caused breaking changes). Here is the repo that has been used perform this test: https://github.com/devversion/cldr-build-comparison

jessicajaniuk commented:
Good news, everyone! The TGP is green. I think this is safe to land.

devversion committed:
build: generate closure locale files using hard-coded list of locales

With the refactoring from a Gulp task to a Bazel too, we tried switching
away from the hard-coded list of locales and aliases for the Closure
Locale file generation. After multiple attempts of landing this, it
turned out that Closure Compiler/Closure Library relies on locale
identifiers CLDR does not capture within it's `availableLocales.json`
or `aliases.json` data.

Closure Library does not use any unknown locale identifiers here. The
locale identifiers can be resolved within CLDR using the bundle lookup
algorithm that is specified as part of CLDR; instead the problem is that
the locale identifiers do not follow any reasonable pattern and
therefore it's extremely difficult to generate them automatically (it's
almost like we'd need to build up _all_ possible combinations). Instead
of doing that, we just use the hard-coded locales and aliases from the
old Closure Locale generation script.

devversion commented:
It seems like I was too optimistic about no longer hard-coding the locales and aliases used by Closure Library for i18n. I've updated the closure locale file generation so that the files matches with the file from the Gulp generation.

I've also re-compared the public locale data, plus this time the closure locale file. Looks like this time we should be really safe (also in terms of g3 now). https://github.com/devversion/cldr-build-comparison/commit/f9480340a5fc733bb20972c10fd2fdb0a42ead47

jessicajaniuk commented:
@devversion looks like this needs a rebase and a finished presubmit.

devversion commented:
@jessicajaniuk thanks for the heads up. Rebased. yeah, I'm waiting for this change to be ready in g3 (we want to manually test the change in an app that previously failed due to this change)

AndrewKushnir commented:
@devversion it looks like there is a merge conflict with the main branch (after merging other changes earlier). Could you please resolve the conflict when you get a chance?

I've also added the "blocked" label for now since we'd need to perform extra testing in Google's codebase before landing the change (to avoid rollbacks).

AndrewKushnir commented:
FYI, removing from the merge queue for now (by removing the "merge" label), since some extra checks should be performed in g3 before we can land this change.
